validator-network/src/network_impl.rs

```diff
@@ -76,50 +88,111 @@ where
+        };
+        self.validator_peer_id_cache
+            .write()
+            .insert(public_key.compressed().clone(), cache_value);
+    }
+
     /// Look up the peer ID for a validator ID.
     async fn get_validator_peer_id(
         &self,
         validator_id: u16,
-    ) -> Result<N::PeerId, NetworkError<N::Error>> {
-        let (peer_id, public_key) = {
-            let public_key = self
-                .validator_keys
-                .read()
-                .get(usize::from(validator_id))
-                .ok_or(NetworkError::UnknownValidator(validator_id))?
-                .clone();
+    ) -> Result<N::PeerId, NetworkError<N::Error>>
+    where
+        <N as Network>::Error: Sync + Send,
+    {
+        let public_key = self
+            .validator_keys
+            .read()
+            .get(usize::from(validator_id))
+            .ok_or(NetworkError::UnknownValidator(validator_id))?
+            .clone();
 
-            let peer_id = self
+        let update_peer_cache = {
+            if let Some(cache_state) = self
                 .validator_peer_id_cache
                 .read()
                 .get(&public_key.compressed().clone())
-                .cloned();
-            (peer_id, public_key)
+            {
+                match cache_state {
+                    CacheState::Resolved(peer_id) => return Ok(*peer_id),
+                    CacheState::Error(_) => true,
+                    CacheState::InProgress => {
+                        log::debug!(validator_id, "Record resolution is in progress");
+                        false
+                    }
+                }
+            } else {
+                log::error!(
+                    ?public_key,
+                    validator_id,
+                    "Could not find peer ID for validator in DHT",
+                );
+                true
+            }
         };
-
-        if let Some(peer_id) = peer_id {
-            Ok(peer_id)
-        } else if let Some(peer_id) = Self::resolve_peer_id(&self.network, &public_key).await? {
-            self.validator_peer_id_cache
-                .write()
-                .insert(public_key.compressed().clone(), peer_id);
-            Ok(peer_id)
-        } else {
-            log::error!(
-                "Could not find peer ID for validator in DHT: public_key = {:?}",
-                public_key
-            );
-            Err(NetworkError::UnknownValidator(validator_id))
+        if update_peer_cache {
+            // Cache is empty for this validator ID, query the entry
+            {
+                self.validator_peer_id_cache
+                    .write()
+                    .insert(public_key.compressed().clone(), CacheState::InProgress);
+            }
+            let self_ = self.arc_clone();
+            tokio::spawn(async move {
+                Self::update_peer_id_cache(&self_, validator_id, &public_key).await;
+            });
         }
+        Err(NetworkError::UnknownValidator(validator_id))
```

Comment:
If I understand this code correctly, it changes the behavior for validators whose peer IDs are not yet known. Previously, the peer ID would be looked up and then the message would be sent. Now, the peer ID will be looked up, but the message will be reported as failed. Is that intentional?

> This also makes the `validator-network` able to quickly return if the DHT query is still in progress allowing upper layers to continue their process without blocking them until the query finishes or errors out.

I guess that sounds like "yes".

Reply:
yes